### PR TITLE
backport lua_rawlen, lua_len & luaL_len from 5.2

### PR DESCRIPTION
**Pull request by fperrad**
initially submitted as https://github.com/LuaJIT/LuaJIT/pull/327

### CODE REVIEW

**Comment by siddhesh** [comment]
Thank you for the patchset.  Can you please also add documentation for the functions to doc/extensions.html?  I think for this change that should be sufficient.

I'd love to get luajit on par with lua as far as API/ABI is concerned.  We can figure out workarounds for ABI breakages once we get to that bridge.

**Comment by fperrad** [comment]
@siddhesh, documentation added

File: src/lj_api.c [hunk 1]
```diff
@@ -578,6 +580,20 @@ LUA_API size_t lua_objlen(lua_State *L, int idx)
   }
 }
 
+LUA_API size_t lua_rawlen(lua_State *L, int idx)
```
Comment (siddhesh) [inline src/lj_api.c:583]: Hmm, this implementation is slightly different from lua_rawlen but that is due to Mike's 'enhancement' to the lua_objlen where he formats numbers into their string format and returns the resultant length.  I'm OK with letting this diverge.

**Review by siddhesh — APPROVED** [review]
There is this one caveat about lua_rawlen vs lua_objlen due to Mike's extension of lua_objlen behaviour, but we are doing a clean break from it and should be OK.

**Comment by siddhesh** [comment]
Merged manually into master.  I've not put it into the v2.1 branch because I'm trying to keep it backward compatible and hence putting in bug fixes only for now.

Thanks!

**Pull request closed without merging**